internal/transport/transport_test.go

```diff
@@ -3204,3 +3206,141 @@ func (s) TestClientTransport_Handle1xxHeaders(t *testing.T) {
+			defer client.Close(fmt.Errorf("test cleanup"))
+
+			// Wait for connection to be established
+			waitWhileTrue(t, func() (bool, error) {
+				server.mu.Lock()
+				defer server.mu.Unlock()
+				if len(server.conns) == 0 {
+					return true, fmt.Errorf("timed-out while waiting for connection")
+				}
+				return false, nil
+			})
+
+			// Get the server transport
+			server.mu.Lock()
+			var serverTransport *http2Server
+			for st := range server.conns {
+				serverTransport = st.(*http2Server)
+				break
+			}
+			server.mu.Unlock()
+
+			if serverTransport == nil {
+				t.Fatal("Server transport not found")
+			}
+
+			// Always create a real stream through the client
+			ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+			defer cancel()
+			clientStream, err := client.NewStream(ctx, &CallHdr{})
+			if err != nil {
+				t.Fatalf("Failed to create stream: %v", err)
+			}
+
+			// Wait for the stream to be created on the server side
+			var serverStream *ServerStream
+			waitWhileTrue(t, func() (bool, error) {
+				serverTransport.mu.Lock()
+				defer serverTransport.mu.Unlock()
+				for _, v := range serverTransport.activeStreams {
+					if v.id == clientStream.id {
+						serverStream = v
+						return false, nil
+					}
+				}
+				return true, nil
+			})
+
+			if serverStream == nil {
+				t.Fatalf("Server stream not found for client stream ID %d", clientStream.id)
+			}
+
+			// For the NonActiveStream test, remove the stream from activeStreams
+			if test.removeFromActive {
+				serverTransport.mu.Lock()
+				delete(serverTransport.activeStreams, serverStream.id)
+				serverTransport.mu.Unlock()
+			}
+
+			initialSucceeded := serverTransport.channelz.SocketMetrics.StreamsSucceeded.Load()
+			initialFailed := serverTransport.channelz.SocketMetrics.StreamsFailed.Load()
+
+			serverTransport.deleteStream(serverStream, test.eosReceived)
+			serverTransport.deleteStream(serverStream, test.eosReceived)
+			serverTransport.deleteStream(serverStream, test.eosReceived)
+
+			finalSucceeded := serverTransport.channelz.SocketMetrics.StreamsSucceeded.Load()
+			finalFailed := serverTransport.channelz.SocketMetrics.StreamsFailed.Load()
+
+			gotSucceededDelta := finalSucceeded - initialSucceeded
+			gotFailedDelta := finalFailed - initialFailed
+
+			if gotSucceededDelta != test.wantStreamSucceeded {
+				t.Errorf("StreamsSucceeded: got delta %d, want %d", gotSucceededDelta, test.wantStreamSucceeded)
+			}
+			if gotFailedDelta != test.wantStreamFailed {
+				t.Errorf("StreamsFailed: got delta %d, want %d", gotFailedDelta, test.wantStreamFailed)
+			}
+
+			// Verify stream was removed from activeStreams (unless it was pre-removed for NonActiveStream test)
+			if !test.removeFromActive {
+				serverTransport.mu.Lock()
+				_, exists := serverTransport.activeStreams[serverStream.id]
+				serverTransport.mu.Unlock()
+				if exists {
+					t.Error("Stream should have been removed from activeStreams")
+				}
+			}
```

Comment:
We can remove the assertions on the `activeStreams` map here. Ideally we should test the public API of a struct (methods on `serverTransport`) instead of it's internal details. In this case, testing the metric should give us enough confidence that the state of the map is correct.

Reply:
[6898146](https://github.com/grpc/grpc-go/pull/8573/commits/6898146e5bea3fc29f671b133c039d6d71b27bf7)